File: mobile/android/focus-android/app/src/main/java/org/mozilla/focus/webview/matcher/UrlMatcher.java
/* -*- Mode: Java; c-basic-offset: 4; tab-width: 20; indent-tabs-mode: nil; -*-
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
package org.mozilla.focus.webview.matcher;


import android.content.Context;
import android.content.SharedPreferences;
import android.net.Uri;
import android.preference.PreferenceManager;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;
import androidx.collection.ArrayMap;
import android.util.JsonReader;

import org.mozilla.focus.R;
import org.mozilla.focus.utils.Settings;
import org.mozilla.focus.webview.matcher.util.FocusString;

import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class UrlMatcher implements  SharedPreferences.OnSharedPreferenceChangeListener {
    /**
     * Map of pref to blocking category (preference key -> Blocklist category name).
     */
    private final Map<String, String> categoryPrefMap;

    private static final String[] WEBFONT_EXTENSIONS = new String[]{
            ".woff2",
            ".woff",
            ".eot",
            ".ttf",
            ".otf"
    };

    private static final String WEBFONTS = "Webfonts";

    private static Map<String, String> loadDefaultPrefMap(final Context context) {
        Map<String, String> tempMap = new ArrayMap<>();

        tempMap.put(context.getString(R.string.pref_key_privacy_block_ads), "Advertising");
        tempMap.put(context.getString(R.string.pref_key_privacy_block_analytics), "Analytics");
        tempMap.put(context.getString(R.string.pref_key_privacy_block_social), "Social");
        tempMap.put(context.getString(R.string.pref_key_privacy_block_other), "Content");
        tempMap.put(context.getString(R.string.pref_key_privacy_block_cryptomining), "Cryptomining");
        tempMap.put(context.getString(R.string.pref_key_privacy_block_fingerprinting), "Fingerprinting");

        // This is a "fake" category - webfont handling is independent of the blocklists
        tempMap.put(context.getString(R.string.pref_key_performance_block_webfonts), WEBFONTS);

        return Collections.unmodifiableMap(tempMap);
    }

    private final Map<String, Trie> categories;
    private final Set<String> enabledCategories = new HashSet<>();

    private final EntityList entityList;
    // A cached list of previously matched URLs. This MUST be cleared whenever items are removed from enabledCategories.
    private final HashSet<String> previouslyMatched = new HashSet<>();
    // A cahced list of previously approved URLs. This MUST be cleared whenever items are added to enabledCategories.
    private final HashSet<String> previouslyUnmatched = new HashSet<>();

    private boolean blockWebfonts = true;

    public static UrlMatcher loadMatcher(final Context context, final int blockListFile, final int[] blockListOverrides, final int entityListFile) {
        final Map<String, String> categoryPrefMap = loadDefaultPrefMap(context);

        final Map<String, Trie> categoryMap = new HashMap<>(5);
        try (final JsonReader jsonReader =
                     new JsonReader(new InputStreamReader(context.getResources().openRawResource(blockListFile), StandardCharsets.UTF_8))) {
            BlocklistProcessor.loadCategoryMap(jsonReader, categoryMap, BlocklistProcessor.ListType.BASE_LIST);
        } catch (IOException e) {
            throw new IllegalStateException("Unable to parse blacklist");
        }

        if (blockListOverrides != null) {
            for (int blockListOverride : blockListOverrides) {
                try (final JsonReader jsonReader =
                             new JsonReader(new InputStreamReader(context.getResources().openRawResource(blockListOverride), StandardCharsets.UTF_8))) {
                    BlocklistProcessor.loadCategoryMap(jsonReader, categoryMap, BlocklistProcessor.ListType.OVERRIDE_LIST);
                } catch (IOException e) {
                    throw new IllegalStateException("Unable to parse override blacklist");
                }
            }
        }

        final EntityList entityList;
        try (final JsonReader jsonReader = new JsonReader(new InputStreamReader(context.getResources().openRawResource(entityListFile), StandardCharsets.UTF_8))) {
            entityList = EntityListProcessor.getEntityMapFromJSON(jsonReader);
        } catch (IOException e) {
            throw new IllegalStateException("Unable to parse entity list");
        }

        return new UrlMatcher(context, categoryPrefMap, categoryMap, entityList);
    }

    /* package-private */ UrlMatcher(final Context context,
                                     @NonNull final Map<String, String> categoryPrefMap,
                                     @NonNull final Map<String, Trie> categoryMap,
                                     @Nullable final EntityList entityList) {
        this.categoryPrefMap = categoryPrefMap;
        this.entityList = entityList;
        this.categories = categoryMap;

        // Ensure all categories have been declared, and enable by default (loadPrefs() will then
        // enabled/disable categories that have actually been configured).
        for (final Map.Entry<String, Trie> entry: categoryMap.entrySet()) {
            if (!categoryPrefMap.values().contains(entry.getKey())) {
                throw new IllegalArgumentException("categoryMap contains undeclared category: " + entry.getKey());
            }

            // Failsafe: enable all categories (we load preferences in the next step anyway)
            enabledCategories.add(entry.getKey());
        }

        loadPrefs(context);

        PreferenceManager.getDefaultSharedPreferences(context).registerOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onSharedPreferenceChanged(final SharedPreferences sharedPreferences, final String prefName) {
        final String categoryName = categoryPrefMap.get(prefName);

        if (categoryName != null) {
            final boolean prefValue = sharedPreferences.getBoolean(prefName, false);

            setCategoryEnabled(categoryName, prefValue);
        }
    }

    private void loadPrefs(final Context context) {
        final SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);

        for (final Map.Entry<String, String> entry : categoryPrefMap.entrySet()) {
            final boolean prefValue;
            if (entry.getKey().equals(context.getString(R.string.pref_key_performance_block_webfonts))) {
                prefValue = Settings.getInstance(context).shouldBlockWebFonts();
            } else if (entry.getKey().equals(context.getString(R.string.pref_key_privacy_block_social))) {
                prefValue = Settings.getInstance(context).shouldBlockSocialTrackers();
            } else if (entry.getKey().equals(context.getString(R.string.pref_key_privacy_block_ads))) {
                prefValue = Settings.getInstance(context).shouldBlockAdTrackers();
            } else if (entry.getKey().equals(context.getString(R.string.pref_key_privacy_block_analytics))) {
                prefValue = Settings.getInstance(context).shouldBlockAnalyticTrackers();
            } else if (entry.getKey().equals(context.getString(R.string.pref_key_privacy_block_other))) {
                prefValue = Settings.getInstance(context).shouldBlockOtherTrackers();
            } else {
                prefValue = prefs.getBoolean(entry.getKey(), true);
            }
            setCategoryEnabled(entry.getValue(), prefValue);
        }
    }

    @VisibleForTesting UrlMatcher(final String[] patterns) {
        final Map<String, String> map = new HashMap<>();
        map.put("default", "default");
        categoryPrefMap = Collections.unmodifiableMap(map);

        categories = new HashMap<>();

        buildMatcher(patterns);

        entityList = null;
    }

    /**
     * Only used for testing - uses a list of urls to populate a "default" category.
     * @param patterns
     */
    private void buildMatcher(String[] patterns) {
        final Trie defaultCategory;
        if (!categories.containsKey("default")) {
            defaultCategory = Trie.createRootNode();
            categories.put("default", defaultCategory);
        } else {
            defaultCategory = categories.get("default");
        }

        for (final String pattern : patterns) {
            defaultCategory.put(FocusString.create(pattern).reverse());
        }

        enabledCategories.add("default");
    }

    public Set<String> getCategories() {
        return categories.keySet();
    }

    public void setCategoryEnabled(final String category, final boolean enabled) {
        if (WEBFONTS.equals(category)) {
            blockWebfonts = enabled;
            return;
        }

        if (!getCategories().contains(category)) {
            throw new IllegalArgumentException("Can't enable/disable inexistant category");
        }

        if (enabled) {
            if (enabledCategories.contains(category)) {
                // Early return - nothing to do if the category is already enabled
            } else {
                enabledCategories.add(category);
                previouslyUnmatched.clear();
            }
        } else {
            if (!enabledCategories.contains(category)) {
                // Early return - nothing to do if the category is already disabled
            } else {
                enabledCategories.remove(category);
                previouslyMatched.clear();
            }

        }
    }

    public boolean matches(final Uri resourceURI, final Uri pageURI) {
        final String path = resourceURI.getPath();

        if (path == null) {
            return false;
        }

        // We need to handle webfonts first: if they are blocked, then whitelists don't matter.
        // If they aren't blocked we still need to check domain blacklists below.
        if (blockWebfonts) {
            for (final String extension : WEBFONT_EXTENSIONS) {
                if (path.endsWith(extension)) {
                    return true;
                }
            }
        }

        final String resourceURLString = resourceURI.toString();

        // Cached whitelisted items can be permitted now (but blacklisted needs to wait for the override / entity list)
        if (previouslyUnmatched.contains(resourceURLString)) {
            return false;
        }

        if (entityList != null &&
                entityList.isWhiteListed(pageURI, resourceURI)) {
            // We must not cache entityList items (and/or if we did, we'd have to clear the cache
            // on every single location change)
            return false;
        }

        final String resourceHost = resourceURI.getHost();
        final String pageHost = pageURI.getHost();

        if (pageHost != null && pageHost.equals(resourceHost)) {
            return false;
        }

        if (previouslyMatched.contains(resourceURLString)) {
            return true;
        }

        final FocusString revhost = FocusString.create(resourceHost).reverse();

        for (final Map.Entry<String, Trie> category : categories.entrySet()) {
            if (enabledCategories.contains(category.getKey()) &&
                    category.getValue().findNode(revhost) != null) {
                previouslyMatched.add(resourceURLString);
                return true;
            }
        }

        previouslyUnmatched.add(resourceURLString);
        return false;
    }
}

File: mobile/android/android-components/components/browser/engine-system/src/main/java/mozilla/components/browser/engine/system/matcher/UrlMatcher.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package mozilla.components.browser.engine.system.matcher

import android.content.Context
import android.content.res.Resources
import android.net.Uri
import android.util.JsonReader
import androidx.annotation.RawRes
import java.io.InputStreamReader
import java.io.Reader
import java.nio.charset.StandardCharsets.UTF_8
import java.util.LinkedList

/**
 * Provides functionality to process categorized URL block/safe lists and match
 * URLs against these lists.
 */
class UrlMatcher {
    private val categories: MutableMap<String, Trie>
    internal val enabledCategories = HashSet<String>()

    private val safelist: Safelist?
    private val previouslyMatched = HashSet<String>()
    private val previouslyUnmatched = HashSet<String>()

    constructor(patterns: Array<String>) {
        categories = HashMap()
        safelist = null

        val defaultCategory = Trie.createRootNode()
        patterns.forEach { defaultCategory.put(it.reverse()) }
        categories[DEFAULT] = defaultCategory
        enabledCategories.add(DEFAULT)
    }

    internal constructor(
        enabledCategories: Set<String>,
        supportedCategories: Set<String>,
        categoryMap: MutableMap<String, Trie>,
        safelist: Safelist? = null,
    ) {
        this.safelist = safelist
        this.categories = categoryMap

        for ((key) in categoryMap) {
            if (!supportedCategories.contains(key)) {
                throw IllegalArgumentException("$key categoryMap contains undeclared category")
            }
        }

        enabledCategories.forEach { setCategoryEnabled(it, true) }
    }

    /**
     * Enables the provided categories.
     *
     * @param categories set of categories to enable.
     */
    fun setCategoriesEnabled(categories: Set<String>) {
        if (enabledCategories != categories) {
            enabledCategories.removeAll { it != DEFAULT }
            categories.forEach { setCategoryEnabled(it, true) }
        }
    }

    internal fun setCategoryEnabled(category: String, enabled: Boolean) {
        if (enabled) {
            if (enabledCategories.contains(category)) {
                return
            } else {
                enabledCategories.add(category)
                previouslyUnmatched.clear()
            }
        } else {
            if (!enabledCategories.contains(category)) {
                return
            } else {
                enabledCategories.remove(category)
                previouslyMatched.clear()
            }
        }
    }

    /**
     * Checks if the given page URI is blocklisted for the given resource URI.
     * Returns true if the site (page URI) is allowed to access
     * the resource URI, otherwise false.
     *
     * @param resourceURI URI of a resource to be loaded by the page
     * @param pageURI URI of the page
     * @return a [Pair] of <Boolean, String?> the first indicates, if the URI matches and the second
     * indicates the category of the match if available otherwise null.
     */
    fun matches(resourceURI: String, pageURI: String): Pair<Boolean, String?> {
        return matches(Uri.parse(resourceURI), Uri.parse(pageURI))
    }

    /**
     * Checks if the given page URI is blocklisted for the given resource URI.
     * Returns true if the site (page URI) is allowed to access
     * the resource URI, otherwise false.
     *
     * @param resourceURI URI of a resource to be loaded by the page
     * @param pageURI URI of the page
     * @return a [Pair] of <Boolean, String?> the first indicates, if the URI matches and the second
     * indicates the category of the match if available otherwise null.
     */
    @Suppress("ReturnCount", "ComplexMethod")
    fun matches(resourceURI: Uri, pageURI: Uri): Pair<Boolean, String?> {
        val resourceURLString = resourceURI.toString()
        val resourceHost = resourceURI.host
        val pageHost = pageURI.host
        val notMatchesFound = false to null

        if (previouslyUnmatched.contains(resourceURLString)) {
            return notMatchesFound
        }

        if (safelist?.contains(pageURI, resourceURI) == true) {
            return notMatchesFound
        }

        if (pageHost != null && pageHost == resourceHost) {
            return notMatchesFound
        }

        if (previouslyMatched.contains(resourceURLString)) {
            return true to null
        }

        if (resourceHost == null) {
            return notMatchesFound
        }

        for ((key, value) in categories) {
            if (enabledCategories.contains(key) && value.findNode(resourceHost.reverse()) != null) {
                previouslyMatched.add(resourceURLString)
                return true to key
            }
        }

        previouslyUnmatched.add(resourceURLString)
        return notMatchesFound
    }

    companion object {
        const val ADVERTISING = "Advertising"
        const val ANALYTICS = "Analytics"
        const val CONTENT = "Content"
        const val SOCIAL = "Social"
        const val DEFAULT = "default"
        const val CRYPTOMINING = "Cryptomining"
        const val FINGERPRINTING = "Fingerprinting"

        private val ignoredCategories = setOf("Legacy Disconnect", "Legacy Content")
        private val webfontExtensions = arrayOf(".woff2", ".woff", ".eot", ".ttf", ".otf")
        private val supportedCategories = setOf(
            ADVERTISING,
            ANALYTICS,
            SOCIAL,
            CONTENT,
            CRYPTOMINING,
            FINGERPRINTING,
        )

        /**
         * Creates a new matcher instance for the provided URL lists.
         *
         * @deprecated Pass resources directly
         * @param blocklistFile resource ID to a JSON file containing the block list
         * @param safelistFile resource ID to a JSON file containing the safe list
         */
        fun createMatcher(
            context: Context,
            @RawRes blocklistFile: Int,
            @RawRes safelistFile: Int,
            enabledCategories: Set<String> = supportedCategories,
        ): UrlMatcher =
            createMatcher(context.resources, blocklistFile, safelistFile, enabledCategories)

        /**
         * Creates a new matcher instance for the provided URL lists.
         *
         * @param blocklistFile resource ID to a JSON file containing the block list
         * @param safelistFile resource ID to a JSON file containing the safe list
         */
        fun createMatcher(
            resources: Resources,
            @RawRes blocklistFile: Int,
            @RawRes safelistFile: Int,
            enabledCategories: Set<String> = supportedCategories,
        ): UrlMatcher {
            val blocklistReader = InputStreamReader(resources.openRawResource(blocklistFile), UTF_8)
            val safelistReader = InputStreamReader(resources.openRawResource(safelistFile), UTF_8)
            return createMatcher(blocklistReader, safelistReader, enabledCategories)
        }

        /**
         * Creates a new matcher instance for the provided URL lists.
         *
         * @param block reader containing the block list
         * @param safe resource ID to a JSON file containing the safe list
         */
        fun createMatcher(
            block: Reader,
            safe: Reader,
            enabledCategories: Set<String> = supportedCategories,
        ): UrlMatcher {
            val categoryMap = HashMap<String, Trie>()

            JsonReader(block).use {
                    jsonReader ->
                loadCategories(jsonReader, categoryMap)
            }

            var safelist: Safelist?
            JsonReader(safe).use { jsonReader -> safelist = Safelist.fromJson(jsonReader) }
            return UrlMatcher(enabledCategories, supportedCategories, categoryMap, safelist)
        }

        /**
         * Checks if the given URI points to a Web font.
         * @param uri the URI to check.
         *
         * @return true if the URI is a Web font, otherwise false.
         */
        fun isWebFont(uri: Uri): Boolean {
            val path = uri.path ?: return false
            return webfontExtensions.find { path.endsWith(it) } != null
        }

        private fun loadCategories(
            reader: JsonReader,
            categoryMap: MutableMap<String, Trie>,
            override: Boolean = false,
        ): Map<String, Trie> {
            reader.beginObject()

            while (reader.hasNext()) {
                val name = reader.nextName()
                if (name == "categories") {
                    extractCategories(reader, categoryMap, override)
                } else {
                    reader.skipValue()
                }
            }

            reader.endObject()
            return categoryMap
        }

        @Suppress("ThrowsCount", "ComplexMethod", "NestedBlockDepth")
        private fun extractCategories(reader: JsonReader, categoryMap: MutableMap<String, Trie>, override: Boolean) {
            reader.beginObject()

            val socialOverrides = LinkedList<String>()
            while (reader.hasNext()) {
                val categoryName = reader.nextName()
                when {
                    ignoredCategories.contains(categoryName) -> reader.skipValue()
                    else -> {
                        val categoryTrie: Trie?
                        if (!override) {
                            if (categoryMap.containsKey(categoryName)) {
                                throw IllegalStateException("Cannot insert already loaded category")
                            }
                            categoryTrie = Trie.createRootNode()
                            categoryMap[categoryName] = categoryTrie
                        } else {
                            categoryTrie = categoryMap[categoryName]
                            if (categoryTrie == null) {
                                throw IllegalStateException("Cannot add override items to nonexistent category")
                            }
                        }
                        extractCategory(reader) { url, _ -> categoryTrie.put(url.reverse()) }
                    }
                }
            }

            val socialTrie = categoryMap[SOCIAL]
            if (socialTrie == null && !override) {
                throw IllegalStateException("Expected social list to exist")
            }
            socialOverrides.forEach { socialTrie?.put(it.reverse()) }
            reader.endObject()
        }

        private fun extractCategory(reader: JsonReader, callback: (String, String) -> Unit) {
            reader.beginArray()
            while (reader.hasNext()) {
                extractSite(reader, callback)
            }
            reader.endArray()
        }

        private fun extractSite(reader: JsonReader, callback: (String, String) -> Unit) {
            reader.beginObject()
            val siteOwner = reader.nextName()

            reader.beginObject()
            while (reader.hasNext()) {
                reader.skipValue()
                val nextToken = reader.peek()
                if (nextToken.name == "STRING") {
                    // Sometimes there's a "dnt" entry, with unspecified purpose.
                    reader.skipValue()
                } else {
                    reader.beginArray()
                    while (reader.hasNext()) {
                        callback(reader.nextString(), siteOwner)
                    }
                    reader.endArray()
                }
            }
            reader.endObject()

            reader.endObject()
        }
    }
}
